external/qjs_fs_module.c

```diff
@@ -0,0 +1,3076 @@
+    JSValueConst *argv, int calltype)
+{
+    ssize_t                      n;
+    JSValue                      callback, v, result, encode, options;
+    njs_str_t                    str;
+    const char                   *path, *enc;
+    const qjs_buffer_encoding_t  *encoding;
+    char                         path_buf[NJS_MAX_PATH + 1],
+                                 dst_buf[NJS_MAX_PATH + 1];
+
+    path = qjs_fs_path(cx, path_buf, argv[0], "path");
+    if (path == NULL) {
+        return JS_EXCEPTION;
+    }
+
+    options = argv[1];
+    encode = JS_UNDEFINED;
+    callback = JS_UNDEFINED;
+
+    if (calltype == QJS_FS_CALLBACK) {
+        callback = argv[njs_min(argc - 1, 2)];
+
+        if (!JS_IsFunction(cx, callback)) {
+            JS_ThrowTypeError(cx, "\"callback\" must be a function");
+            return JS_EXCEPTION;
+        }
+
+        if (JS_SameValue(cx, options, callback)) {
+            options = JS_UNDEFINED;
+        }
+    }
+
+    if (JS_IsString(options)) {
+        encode = JS_DupValue(cx, options);
+
+    } else if (!JS_IsUndefined(options)) {
+        if (!JS_IsObject(options)) {
+            JS_ThrowTypeError(cx, "Unknown options type (a string or object "
+                              "required)");
+            return JS_EXCEPTION;
+        }
+
+        v = JS_GetPropertyStr(cx, options, "encoding");
+        if (!JS_IsUndefined(v) && !JS_IsException(v)) {
+            encode = v;
+        }
+    }
+
+    encoding = NULL;
+    enc = JS_ToCString(cx, encode);
+    if (enc == NULL) {
+        JS_FreeValue(cx, encode);
+        return JS_EXCEPTION;
+    }
+
+    if (strncmp(enc, "buffer", 6) != 0) {
+        encoding = qjs_buffer_encoding(cx, encode, 1);
+        if (encoding == NULL) {
+            JS_FreeCString(cx, enc);
+            JS_FreeValue(cx, encode);
+            return JS_EXCEPTION;
+        }
+    }
+
+    JS_FreeCString(cx, enc);
+    JS_FreeValue(cx, encode);
+
+    str.start = (u_char *) dst_buf;
+    n = readlink(path, dst_buf, sizeof(dst_buf) - 1);
+    if (n < 0) {
+        result = qjs_fs_error(cx, "readlink", strerror(errno), path, errno);
+        goto done;
+    }
+
+    str.length = n;
+
+    result = qjs_fs_encode(cx, encoding, &str);
+
+done:
+
+    if (JS_IsException(result)) {
+        return JS_EXCEPTION;
+    }
+
+    return qjs_fs_result(cx, result, calltype, callback);
+}
+
+
+static JSValue
+qjs_fs_dirent_create(JSContext *cx, JSValueConst name, struct dirent *entry)
+{
+    JSValue  obj, type;
+
+    obj = JS_NewObjectClass(cx, QJS_CORE_CLASS_ID_FS_DIRENT);
+    if (JS_IsException(obj)) {
+        return JS_EXCEPTION;
+    }
+
+    if (JS_DefinePropertyValueStr(cx, obj, "name", name,
+                                  JS_PROP_ENUMERABLE) < 0)
+    {
+        JS_FreeValue(cx, obj);
+        return JS_EXCEPTION;
+    }
+
+    if (entry != NULL) {
+        type = JS_NewInt32(cx, qjs_dentry_type(entry));
+        if (JS_IsException(type)) {
```

Comment:
JS_NewInt32 never produce exception, so no need test for it here and in other places.